CIP-XXXX/README.md

```diff
@@ -0,0 +1,293 @@
+
+2. **Using general-purpose query layers**. General-purpose query layers allow to query blockchain data using a wide set of APIs that are not built with a particular dApp domain in mind. dApp state has to be constructed based on data returned from the queries. Examples: Blockfrost, Maestro.
+
+The first approach allows for lower runtime resource consumption, but a general-purpose query layer has an advantage of being more easily reusable between dApps.
+
+In this proposal, we are focusing on general-purpose querying only.
+
+#### Handling of rollbacks
+
+Transaction rollbacks are essential to blockchains: local node's view of the chain may be different from other nodes'. During conflict resolution, the node may issue a rollback event, that should be handled by dApps.
+
+Customized chain followers, at least in principle, allow for "live" rollback handling: that is, a user-facing dApp can subscribe to a local view of a part of the UTxO set.
+
+General purpose query layers can also handle rollbacks just fine, but they don't propagate rollback events to dApps, because they do not possess any dApp-specific info to determine if a dApp *needs* to handle a particular rollback. dApps that work with general-purpose query layers follow pull-based architecture, rather than event subscription-based, which means they just request data as needed, instead of reacting to blockchain events.
+
+In the context of this API, rollbacks should be acknowledged as a source of potential inconsistency between data pieces returned by different queries.
+
+#### Error handling
+
+Errors should be divided in two categories:
+
+- domain errors
+- transport errors (404, 500, etc)
+
+This document should only cover domain errors.
+
+#### Pagination
+
+In CIP-30, pagination is not reliable, because there is no guarantee that the set of UTxOs does not change between calls. This behavior is not suitable for DeFi: consistency should be prioritized, even though it's impossible to guarantee it (due to rollbacks).
+
+Stable pagination can be achieved by letting the user refer to a particular state in time from the recent past via a slot number.
+
+### Methods
+
+This section contains transport-agnostic method descriptions & their parameter lists.
+
+The scope of this section is loosely based on a [comparison table for existing Cardano query layers](./Query_Layer_API_Comparison.md).
+The goal is to make it so that the API could be implemented via simple adapters that transform requests and responses to the appropriate formats.
+
+The payload formats used below are references to [CIP-0116 - Standard JSON encoding for Domain Types](https://cips.cardano.org/cip/CIP-0116), which specifies cardano domain types via a JSON schema.
+
+#### UTxOs
+
+##### By address
+
+Request:
+
+- address: `Address`
+- slot?: `UInt64`
+
+Response:
+
+- `TransactionUnspentOutput[]`
```

Comment:
CIP-0116 defines this to include a `TransactionOutput` with this schema: https://github.com/cardano-foundation/CIPs/blob/b81611632f9dcea0b87d7d96cf673a720c77e929/CIP-0116/cardano-babbage.json#L445C6-L445C23

I would like to point out that chain data APIs seemingly do not decode the datum by default, i.e. when responding with transaction outputs.

I just stumbled over this yesterday [in Hydra](https://github.com/cardano-scaling/hydra/issues/1543), which uses the `cardano-api` library to encode `TxOut`s and that results in a decoded datum mapped to json equivalence. See also our (messy, sorry) api reference: https://hydra.family/head-protocol/unstable/api-reference/#schema-TxOut